lib/web_ui/lib/src/engine/canvaskit/image.dart

```diff
@@ -4,24 +4,200 @@
+
+  if (browserSupportsImageDecoder) {
+    codec = await CkBrowserImageDecoder.create(
       data: list,
+      contentType: contentType,
       debugSource: 'encoded image bytes',
     );
   } else {
-    return CkAnimatedImage.decodeFromBytes(list, 'encoded image bytes', targetWidth: targetWidth, targetHeight: targetHeight);
+    // TODO(harryterkelsen): If the image is animated, then use Skia to decode.
+    // This is currently too conservative, assuming all GIF and WEBP images are
+    // animated. We should detect if they are actually animated by reading the
+    // image headers.
+    if (contentType == 'image/gif' || contentType == 'image/webp') {
+      codec = CkAnimatedImage.decodeFromBytes(list, 'decoded bytes',
+          targetWidth: targetWidth, targetHeight: targetHeight);
+    } else {
+      final DomBlob blob = createDomBlob(<ByteBuffer>[list.buffer]);
+      codec = await decodeBlobToCkImage(blob);
+    }
   }
+  return CkResizingCodec(
+    codec,
+    targetWidth: targetWidth,
+    targetHeight: targetHeight,
+    allowUpscaling: allowUpscaling,
+  );
+}
+
+/// A resizing codec which uses an HTML <canvas> element to scale the image if
+/// it is backed by an HTML Image element.
+class CkResizingCodec extends ResizingCodec {
+  CkResizingCodec(
+    super.delegate, {
+    super.targetWidth,
+    super.targetHeight,
+    super.allowUpscaling,
+  });
+
+  @override
+  ui.Image scaleImage(
+    ui.Image image, {
+    int? targetWidth,
+    int? targetHeight,
+    bool allowUpscaling = true,
+  }) {
+    final CkImage ckImage = image as CkImage;
+    if (ckImage.imageElement == null) {
+      return scaleImageIfNeeded(
+        image,
+        targetWidth: targetWidth,
+        targetHeight: targetHeight,
+        allowUpscaling: allowUpscaling,
+      );
+    } else {
+      return _scaleImageUsingDomCanvas(
+        ckImage,
+        targetWidth: targetWidth,
+        targetHeight: targetHeight,
+        allowUpscaling: allowUpscaling,
+      );
+    }
+  }
+
+  CkImage _scaleImageUsingDomCanvas(
+    CkImage image, {
+    int? targetWidth,
+    int? targetHeight,
+    bool allowUpscaling = true,
+  }) {
+    assert(image.imageElement != null);
+    final int width = image.width;
+    final int height = image.height;
+    final ui.Size? scaledSize =
+        scaledImageSize(width, height, targetWidth, targetHeight);
+    if (scaledSize == null) {
+      return image;
+    }
+    if (!allowUpscaling &&
+        (scaledSize.width > width || scaledSize.height > height)) {
+      return image;
+    }
+
+    final int scaledWidth = scaledSize.width.toInt();
+    final int scaledHeight = scaledSize.height.toInt();
+
+    final DomCanvasElement htmlCanvas = createDomCanvasElement(
+      width: scaledWidth,
+      height: scaledHeight,
+    );
+    final DomCanvasRenderingContext2D ctx =
+        htmlCanvas.getContext('2d')! as DomCanvasRenderingContext2D;
```

Comment:
Unreliable how? I think we should try to avoid explicitly pulling the bytes out of the image. Is there a reason we can't just create an `ImageBitmap` and use that, instead of grabbing the image data? I also should mention that `createImageBitmap` actually allows you to scale the image while extracting it from the image source: https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap